lib/matplotlib/cbook.py

```diff
@@ -492,17 +504,23 @@ def connect(self, s, func):
         func will be called
         """
         self._func_cid_map.setdefault(s, WeakKeyDictionary())
-        if func in self._func_cid_map[s]:
```

Comment:
Good question!  Tl;dr having spent the day on it I still don't know.  And to clarify in case of misunderstanding, by "fail" I mean "will always give ``False``".

I traced the problem as far as ``weakref.ref`` not working for bound or unbound methods, i.e. ``ref(object.method)() == None`` and ``ref(class.method)() == None`` noting that a ``WeakKeyDictionary`` wraps the keys like ``ref(key)``; also when comparing the ``WeakKeyDictionary`` automatically wraps in ref for comparing, as we deal with dead refs, this if statement, ``if ref(func) == ref(func)`` behaves like ``if np.nan == np.nan``, and thus it will always give ``False`` for bound/unbound methods.

Having got so far as discovering the problem lay in ``weakref.ref`` and giving up on figuring out a way to get the function back for comparison, it dawned on me to ask the question why did the callbacks still fire, albeit more then the expected once, and from there the solution unravelled.